### PR TITLE
social authentication

### LINKED ISSUE #153: Implement Google Authentication
First round of improved authentication, aimed at staff since they all have an account already.

### PR DESCRIPTION
**Pull request by Pomax**
fixes #153 aaaaaand just works.

basic admin on http://localhost:5000/admin is still the same, gauth login in on http://localhost:5000/soc/login/google-oauth2/?next=/admin so the admin template still needs an update to have a `or click here to log in using google auth`

This does require having a google auth key and secret of course, and as a thing that's not all too well documented, the gauth key will end in a substring that starts with `.app.` and that part should be removed for the auth key env var.

### CODE REVIEW

File: app/networkapi/settings.py [hunk 1]
```diff
@@ -131,15 +146,19 @@
         ],
         'APP_DIRS': True,
         'OPTIONS': {
-            'context_processors': [
+            'context_processors': list(filter(None, [
+                'social_django.context_processors.backends'
+                if SOCIAL_SIGNIN else None,
+                'social_django.context_processors.login_redirect'
+                if SOCIAL_SIGNIN else None,
```
Comment (cadecairos) [inline app/networkapi/settings.py]: this was hard to parse in my head - I'm guessing it's on a new line because of length... Would some extra indendation make it easier to read, while still keeping lint happy?

File: app/networkapi/settings.py [hunk 2]
```diff
@@ -53,11 +55,18 @@
...
+# Use social authentication if there are key/secret values defined
+SOCIAL_AUTH_GOOGLE_OAUTH2_KEY = env('SOCIAL_AUTH_GOOGLE_OAUTH2_KEY')
+SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET = env('SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET')
+SOCIAL_SIGNIN = SOCIAL_AUTH_GOOGLE_OAUTH2_KEY is not None and \
+                    SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET is not None
+
 # Application definition
 
-INSTALLED_APPS = [
+INSTALLED_APPS = list(filter(None, [
```
Comment (cadecairos) [inline app/networkapi/settings.py]: I'm guessing you do this so that the conditional on 69 can execute?

File: app/networkapi/templates/admin/login.html [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+{% endif %}
+{% if no_site_permission %}
+<p class="errornote">{% trans "You don't have permission to access the admin for this site." %}</p>
+{% endif %}
+
+
+<form action="{{ app_path }}" method="post" id="login-form">
+	<div class="social-login">
+		<p>
+			Log in using <a href="{% url "social:begin" "google-oauth2" %}?next=/admin">Google Authentication</a>.
```
Comment (cadecairos) [inline app/networkapi/templates/admin/login.html]: Can we just say "Log in using Google"?

File: app/networkapi/templates/admin/login.html [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+{% if no_site_permission %}
+<p class="errornote">{% trans "You don't have permission to access the admin for this site." %}</p>
+{% endif %}
+
+
+<form action="{{ app_path }}" method="post" id="login-form">
+    {% if social %}
+    <div class="social-login">
+        <p>
+        Log in using <a href="{% url "social:begin" "google-oauth2" %}?next=/admin">Google Authentication</a>.
```
Comment (cadecairos) [inline app/networkapi/templates/admin/login.html]: can we just say "Log in with Google"?

File: app/networkapi/settings.py (hunk 2 above)
Comment (Pomax) [inline app/networkapi/settings.py]: Correct - the conditional may yield a `None`, which needs to be filtered out because Django will crash hard if there are empty strings or `None` entries in here.

File: app/networkapi/settings.py (hunk 1 above)
Comment (Pomax) [inline app/networkapi/settings.py]: nope, extra indent is a flake8 error
